config.org
#+TITLE: Doom Emacs configuration
#+AUTHOR: Laurent Lejeune
#+HTML_HEAD: <link rel="stylesheet" type="text/css" href="../../org/styles/org.css"/>

* Introduction

This uses the [[https://github.com/doomemacs/doomemacs][doom-emacs]] configuration framework which provides sane defaults,
hence this personal config file is pretty minimal in itself.

* General Bindings

#+begin_src emacs-lisp :tangle yes
(map! :leader
    :desc "Kill buffer" "x" #'iflipb-kill-buffer
    :desc "Switch buffer" ">" #'switch-to-buffer
    :desc "Expand region" "e" #'er/expand-region
    :desc "Increase font size" "+" #'doom/increase-font-size
    :desc "Decrease font size" "-" #'doom/decrease-font-size
    :desc "Next buffer" "k" #'iflipb-next-buffer
    :desc "Previous buffer" "j" #'iflipb-previous-buffer
    :desc "Search buffer" "/" #'+default/search-buffer
    :desc "Search project" "ps" #'+default/search-project
    :desc "Switch window" "TAB" #'ace-window
    :desc "Open URL" "gx" #'browse-url-at-point)
#+end_src

Evil-mode does not conform precisely to vim's standard keybinding.
We reset these here:
#+begin_src emacs-lisp :tangle yes
(with-eval-after-load 'evil
  (define-key evil-normal-state-map (kbd "C-a") 'evil-numbers/inc-at-pt)
  (define-key evil-normal-state-map (kbd "C-x") 'evil-numbers/dec-at-pt)

  (define-key evil-normal-state-map (kbd "C-l") 'evil-window-right)
  (define-key evil-normal-state-map (kbd "C-h") 'evil-window-left)

  (define-key evil-insert-state-map (kbd "C-l") 'evil-window-right)
  (define-key evil-insert-state-map (kbd "C-h") 'evil-window-left)
  (define-key evil-normal-state-map (kbd "C-;") 'embark-act)
  (define-key evil-insert-state-map (kbd "C-;") 'embark-act)
  )
#+end_src

* Treesitter

Add paths to treesitter grammar libraries:
#+begin_src emacs-lisp :tangle yes
(setq treesit-extra-load-path
      (list (concat (getenv "HOME")
                    "/.nix-profile/lib")))
#+end_src

* Wayland
This dynamically sets the ~WAYLAND_DISPLAY~ variable, which is needed
to launch GUI applications.
In practice, this is only needed when using the daemon.

We register our function as a hook for each new emacs frame.

#+begin_src emacs-lisp :tangle yes
(defun ll/set-wayland-socket ()
  (unless (getenv "WAYLAND_DISPLAY")
    (let* ((runtime (or (getenv "XDG_RUNTIME_DIR")
                        (format "/run/user/%s" (user-uid))))
           (wayland (car (directory-files runtime nil "^wayland-[0-9]+$"))))
      (when wayland
        (setenv "WAYLAND_DISPLAY" wayland)))))

(add-hook! 'server-after-make-frame-hook :append #'ll/set-wayland-socket)
#+end_src

* Window Management

#+begin_src emacs-lisp :tangle yes
(defun toggle-maximize-window ()
 "Maximize the window or bring back the previous layout."
  (interactive)
  (if (one-window-p)
    (winner-undo)
    (delete-other-windows)))
#+end_src

#+begin_src emacs-lisp :tangle yes
(define-key evil-normal-state-map (kbd "C-SPC") 'toggle-maximize-window)
#+end_src

* Dired
#+begin_src emacs-lisp :tangle yes
(after! dired
  (dirvish-override-dired-mode t))

(map! :leader
       (:desc "Open" "ff" #'dirvish))
#+end_src
* Drag stuff
#+begin_src emacs-lisp :tangle yes
(use-package! iflipb
  :ensure t
  )
#+end_src

* Drag stuff
Function that allow to move around selections:
#+begin_src emacs-lisp :tangle yes
(define-key evil-visual-state-map (kbd "K") 'drag-stuff-up)
(define-key evil-visual-state-map (kbd "J") 'drag-stuff-down)
#+end_src

* Appearance
** Colors and fonts
#+begin_src emacs-lisp :tangle yes
(use-package! modus-catppuccin
  :ensure t
  :config
  (load-theme 'catppuccin-macchiato :no-confirm))
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 20 :height 1.0 :weight 'normal))
(setq truncate-lines 'nil)
#+end_src

** Indentation guides
#+begin_src emacs-lisp :tangle yes
(setq highlight-indent-guides-auto-odd-face-perc 50)
(setq highlight-indent-guides-auto-even-face-perc 50)
(setq highlight-indent-guides-auto-character-face-perc 50)
(setq highlight-indent-guides-responsive 'top)

(setq display-line-numbers-type 'relative)
#+end_src

Hide emphasis markups mark.
#+begin_src emacs-lisp :tangle yes
(setq org-hide-emphasis-markers t)
#+end_src

** Org-appear

Automatically show/hide org elements when cursor is
placed on them, e.g. will hide/show links.

#+begin_src emacs-lisp :tangle yes
(use-package! org-appear
  :defer t
  :config
  (setq org-appear-autolinks t))
(add-hook! org-mode :append #'org-appear-mode)
#+end_src

* Company

#+begin_src emacs-lisp :tangle yes
(use-package! company
  :defer t
  :custom-face
  (company-tooltip
   ((t (:family "JetBrainsMono Nerd Font" :height 1.0))))
  :config
        (define-key evil-insert-state-map (kbd "C-y") 'company-complete)
        (define-key company-active-map (kbd "RET") nil)
        (define-key company-active-map (kbd "<return>") nil)
        (setq company-idle-delay 0.3)
        (setq company-tooltip-idle-delay 0.3)
        (add-to-list '+lsp-company-backends 'company-files)
        (set-company-backend! 'org-mode 'company-files 'company-capf)
)
#+end_src

* LaTeX
Bind main AucTex menu:

#+begin_src emacs-lisp :tangle yes
(map! :localleader
      :map LaTeX-mode-map
      :desc "Master"    "m" 'TeX-command-master)
#+end_src

Setup default compilation engine.

- The first element of the list is the symbol that AucTeX recognizes.
- The second element is a string with the name of the TeX distribution.
- The third element is the shell command for compiling plain TeX documents.
- The fourth element is the shell command for compiling LaTeX documents.
   Here we are assuming the user is using a Tectonic project (generated using ~tectonic -X new <proj-name>~).
- The last element is the shell command for compiling ConTeXt documents, left unconfigured for now.

#+begin_src emacs-lisp :tangle yes
(setq TeX-engine-alist '((default
                          "Tectonic"
                          "tectonic -X compile -f plain %T"
                          "tectonic -X watch"
                          nil)))
#+end_src

Next, modify the ~LaTeX-command-style~ so that AucTex doesn’t add extra options to it that Tectonic does not recognize.
We simply set it to the ~%(latex)~ expansion (from ~TeX-expand-list-builtin~), removing any other extra options.

#+begin_src emacs-lisp :tangle yes
(setq LaTeX-command-style '(("" "%(latex)")))
#+end_src

We need to set the ~TeX-check-TeX~ variable to ~nil~ since AucTeX will try to find a traditional distribution like ~TeXLive~ or others, and will fail since Tectonic doesn’t meet its criteria.

Additionally, we should also set ~TeX-process-asynchronous~ to ~t~, so that running Tectonic in watch mode doesn’t hang up Emacs.
We’ll also just ensure that the ~TeX-engine~ is set to ~default~.

#+begin_src emacs-lisp :tangle yes
(setq TeX-process-asynchronous t
      TeX-check-TeX nil
      TeX-engine 'default)
#+end_src

Finally, modify the ~TeX-command-list~ to use the appropriate commands and not pass in extra metadata and options to Tectonic that cause it to error out. This needs to be done in place.

#+begin_src emacs-lisp :tangle yes
(add-hook! LaTeX-mode
(let ((tex-list (assoc "TeX" TeX-command-list))
      (latex-list (assoc "LaTeX" TeX-command-list)))
  (setf (cadr tex-list) "%(tex)"
        (cadr latex-list) "%l")))
#+end_src

* Typopunct

#+begin_src emacs-lisp :tangle yes
(use-package! typopunct
  :defer t
  :config
  (typopunct-change-language 'french t))
  #+end_src

* Languagetool
Download standalone [[https://languagetool.org/][LanguageTool]] server from https://languagetool.org/download/LanguageTool-stable.zip and unzip, then set these variables:

#+begin_src emacs-lisp :tangle yes
(setq langtool-language-tool-jar (concat (getenv "HOME") "/bin/LanguageTool-5.2/languagetool-commandline.jar"))

(defun langtool-autoshow-detail-popup (overlays)
  (when (require 'popup nil t)
    ;; Do not interrupt current popup
    (unless (or popup-instances
                ;; suppress popup after type `C-g` .
                (memq last-command '(keyboard-quit)))
      (let ((msg (langtool-details-error-message overlays)))
        (popup-tip msg)))))
(setq langtool-autoshow-message-function
      'langtool-autoshow-detail-popup)

#+end_src

* YAML
#+begin_src emacs-lisp :tangle yes
(use-package! yaml-mode
  :defer t
  :mode ("\\.ya?ml\\'" . yaml-mode))
#+end_src

* Python
** Interpreter
#+begin_src emacs-lisp :tangle yes
(setq python-shell-interpreter "ipython"
    python-shell-interpreter-args "-i --simple-prompt")
#+end_src

#+begin_src emacs-lisp :tangle yes
(map!
 :after python
 :map inferior-python-mode-map
 :n "C-n" #'comint-next-input
 :n "C-p" #'comint-previous-input
 :i "C-n" #'comint-next-input
 :i "C-p" #'comint-previous-input)
#+end_src


** Breakpoints

Bindings and functions to add and highlight breakpoints.
#+begin_src emacs-lisp :tangle yes
(defvar python--pdb-breakpoint-string "breakpoint()")
(defun python-add-breakpoint ()
"Inserts a python breakpoint using `pdb'"
    (interactive)
    (back-to-indentation)
    ;; this preserves the correct indentation in case the line above
    ;; point is a nested block
    (split-line)
    (insert python--pdb-breakpoint-string)
    (python-set-debug-highlight))

(defun python-set-debug-highlight ()
    (interactive)
    (highlight-lines-matching-regexp "breakpoint[.]?" 'hi-red-b))

(defun python-add-debug-highlight ()
    "Adds a highlighter for use by `python--pdb-breakpoint-string'"
    (interactive)
    (python-set-debug-highlight))
    (add-hook 'python-ts-mode-hook 'python-add-debug-highlight)

(map! :localleader
      :map python-ts-mode-map
      :desc "Insert breakpoint"    "d" 'python-add-breakpoint)
#+end_src

** Docstrings

Generate numpy-style docstring automatically from function definition
with [[https://github.com/douglasdavis/numpydoc.el][numpydoc.el]].

#+begin_src emacs-lisp :tangle yes
(use-package! numpydoc
  :defer t
  :init
  (setq numpydoc-insertion-style 'yas))

(map! :localleader
      :map python-ts-mode-map
      :desc "Auto docstring"  "s" 'numpydoc-generate)
#+end_src

* Nix
#+begin_src emacs-lisp :tangle yes
(after! lsp-mode
  (lsp-register-client (make-lsp-client
                        :new-connection (lsp-stdio-connection "nixd")
                        :major-modes '(nix-mode)
                        :server-id 'nixd
                        :priority 10)))
#+end_src

* Docker
Disable format-on-save in docker-mode.
This is acting funny when dealing with ~FROM~ instructions.
#+begin_src emacs-lisp :tangle yes
(setq +format-on-save-disabled-modes (add-to-list '+format-on-save-disabled-modes 'dockerfile-mode))
#+end_src

* Golang
#+begin_src emacs-lisp :tangle yes
(setq lsp-go-use-gofumpt t)
#+end_src

* Scheme

#+begin_src emacs-lisp :tangle yes
(map!
 :after geiser-repl
 :map geiser-repl-mode-map
 :n "C-n" #'comint-next-input
 :n "C-p" #'comint-previous-input
 :i "C-n" #'comint-next-input
 :i "C-p" #'comint-previous-input)
#+end_src
* C++
Enable LSP-based formatting in c++-ts-mode
#+begin_src emacs-lisp :tangle yes
(add-hook 'c++-ts-mode-hook #'+format-with-lsp-mode)
#+end_src

* Harpoon
Quick dynamic bookmarks.

#+begin_src emacs-lisp :tangle yes
(map! :leader
      (:prefix-map ("r" . "Harpoon")
       (:desc "Menu" "m" #'harpoon-quick-menu-hydra
        :desc "Add file" "a" #'harpoon-add-file
        :desc "Edit file" "r" #'harpoon-toggle-file
        :desc "Clear" "c" 'harpoon-clear)))

(map! :leader "1" 'harpoon-go-to-1)
(map! :leader "2" 'harpoon-go-to-2)
(map! :leader "3" 'harpoon-go-to-3)
(map! :leader "4" 'harpoon-go-to-4)
(map! :leader "5" 'harpoon-go-to-5)
(map! :leader "6" 'harpoon-go-to-6)
(map! :leader "7" 'harpoon-go-to-7)
(map! :leader "8" 'harpoon-go-to-8)
(map! :leader "9" 'harpoon-go-to-9)
(map! :leader "0" 'harpoon-go-to-10)
#+end_src
* Org

#+begin_src emacs-lisp :tangle yes
(setq org-export-use-babel t)
(setq org-directory "~/org/")
(setq org-latex-pdf-process '("tectonic %f"))
(setq org-export-in-background t)
#+end_src

#+begin_src emacs-lisp :tangle yes
(use-package! org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config
  (setq org-auto-tangle-default t))
#+end_src

#+begin_src emacs-lisp :tangle yes
(map! :localleader
      :map org-mode-map
      :desc "Preview LaTeX" "u" 'org-latex-preview)
#+end_src

** Backends/Exporters

A couple custom LaTeX classes.

#+begin_src emacs-lisp :tangle yes
(after! ox-latex
    (add-to-list 'org-latex-classes
                '("koma-article" "\\documentclass{scrartcl}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

    (add-to-list 'org-latex-classes
                '("koma-article-fr" "\\documentclass[french]{scrartcl}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

    (add-to-list 'org-latex-classes
                '("memoir-fr"
                "\\documentclass[a4paper,11pt,titlepage, twoside]{memoir}
                    \\usepackage[utf8]{inputenc}
                    \\usepackage[T1]{fontenc}
                    \\usepackage{fixltx2e}
                    \\usepackage{hyperref}
                    \\usepackage{mathpazo}
                    \\usepackage{color}
                    \\usepackage{enumerate}
                    \\definecolor{bg}{rgb}{0.95,0.95,0.95}
                    \\tolerance=1000
                    \\linespread{1.1}
                    \\hypersetup{pdfborder=0 0 0}"
                ("\\chapter{%s}" . "\\chapter*{%s}")
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

    (add-to-list 'org-latex-classes
                '("TMI"
                "\\documentclass[journal, web, twoside]{ieeecolor}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))))
#+end_src

Detect french-style quotes when exporting

#+begin_src emacs-lisp :tangle no
(after! ox
    (setq fr-quotes '("fr"
                    (primary-opening :utf-8 "« " :html "&laquo;&nbsp;" :latex "\\enquote{" :texinfo "@guillemetleft{}@tie{}")
                    (primary-closing :utf-8 " »" :html "&nbsp;&raquo;" :latex "}" :texinfo "@tie{}@guillemetright{}")
                    (secondary-opening :utf-8 "« " :html "&laquo;&nbsp;" :latex "\\\enquote{" :texinfo "@guillemetleft{}@tie{}")
                    (secondary-closing :utf-8 " »" :html "&nbsp;&raquo;" :latex "\\}" :texinfo "@tie{}@guillemetright{}")
                    (apostrophe :utf-8 "’" :html "&rsquo;")))
    (add-to-list 'org-export-smart-quotes-alist fr-quotes))
#+end_src

Exporter for the [[https://github.com/posquit0/Awesome-CV][awesome-cv]] latex class.
#+begin_src emacs-lisp :tangle yes
(use-package! ox-awesomecv
  :after org)
#+end_src

When exporting to HTML, force code block background to use current theme color, otherwise,
things may end up not readable when using dark themes.

#+begin_src emacs-lisp :tangle yes
(defun my/org-inline-css-hook (exporter)
  "Insert custom inline css to automatically set the
background of code to whatever theme I'm using's background"
  (when (eq exporter 'html)
    (let* ((my-pre-bg (face-background 'default))
           (my-pre-fg (face-foreground 'default)))
      (setq
       org-html-head-extra
       (concat
        org-html-head-extra
        (format "<style type=\"text/css\">\n pre.src {background-color: %s; color: %s;}</style>\n"
                my-pre-bg my-pre-fg))))))

(add-hook 'org-export-before-processing-hook 'my/org-inline-css-hook)
#+end_src

** Bibliography Management
Use [[https://github.com/emacs-citar/citar?tab=readme-ov-file#usage][citar]] to auto complete, store notes, and manage a ~bibtex~ bibliographical collection.

#+begin_src emacs-lisp :tangle yes
(use-package! citar
  :config
  (setq citar-bibliography (directory-files-recursively "/home/laurent/org/bibliography/bibs" (concat "\\" ".bib" "$")))
  (setq citar-library-paths '("/home/laurent/org/bibliography/files/")
        citar-notes-paths '("/home/laurent/org/bibliography/notes/"))
  (setq citar-file-open-functions '(("pdf" . citar-file-open-external)
                                    ("html" . citar-file-open-external)
                                    (t . find-file))))

#+end_src

~citar~ is an interface to ~org-cite~ that allows to (among other things):

- Open the ~pdf~ file that correspond to a bibliograph entry (if it exists).
   The file must be located at ~citar-library-paths~ and
  named ~${citekey}.pdf~ to be detected.
- Create an org-roam note using a pre-defined template.
  Here, we generate an ~org-roam~ entry (file) in ~citar-notes-paths~.
- Open the URL of the paper (if it exists).

*** Defining template
Here, we define a number of templates to add a note for each paper, namely:
- ~extended~: This will create a note in a separate file. This
  is the default behaviour when selecting the /Create Org-Roam Notes/ function
  in the main ~citar~ menu.
- ~short~: This will capture a note and append it to an existing org file.
  This boils down to generating a heading and appending it to an existing file.
  This behaviour can be accessed using the custom interactive function defined
  in a next section.

#+begin_src emacs-lisp :tangle yes
(setq org-roam-capture-templates
        '(("s" "short bibliography note" entry "* ${citar-title}\n:PROPERTIES:\n:ID: %(org-id-new)\n:CITEKEY: ${citar-citekey}\n:AUTHOR: ${citar-author}\n:DATE: ${citar-date}\n:END:\n%?"
           :target (file "/home/laurent/org/bibliography/inbox.org")
           :prepend t
           :unnarrowed t
           )
          ("e" "extended bibliography note (separate file)" plain "%?"
           :if-new (file+head "/home/laurent/org/bibliography/notes/${citar-citekey}.org"
                              "#+title: Notes on ${title}#+SETUPFILE: /home/laurent/org/bibliography/setup_file.org\n%?* References :ignore:\n#+print_bibliography:")
           :unnarrowed t)
          ))
#+end_src

We set the "extended" template by default.
#+begin_src emacs-lisp :tangle yes
(setq citar-org-roam-capture-template-key "e")
#+end_src

*** Note taking using interactive template selection

We define an intermediate hydra to select a template prior to creating a note.

#+begin_src emacs-lisp :tangle yes
(defhydra ll/citar-template-hydra (:color blue :hint nil)
  "
Select a template:

_s_: short note
_e_: extended note (separate file)
_q_: quit
"
  ("s"
   (let ((citar-org-roam-capture-template-key "s"))
     (call-interactively #'citar-create-note))
   "short")

  ("e"
   (let ((citar-org-roam-capture-template-key "e"))
     (call-interactively #'citar-create-note))
   "extended")

  ("q" nil "quit"))

(defun ll/citar-create-note ()
  "Prompt for a citar template via Hydra and create a note."
  (interactive)
  (ll/citar-template-hydra/body))
#+end_src

*** Interactively open bibliography file

config.el
(map! :leader
    :desc "Kill buffer" "x" #'kill-current-buffer
    :desc "Switch buffer" ">" #'switch-to-buffer
    :desc "Expand region" "e" #'er/expand-region
    :desc "Increase font size" "+" #'doom/increase-font-size
    :desc "Decrease font size" "-" #'doom/decrease-font-size
    :desc "Next buffer" "k" #'next-buffer
    :desc "Previous buffer" "j" #'previous-buffer
    :desc "Search buffer" "/" #'+default/search-buffer
    :desc "Search buffer" "ps" #'+default/search-project
    :desc "Switch window" "TAB" #'ace-window
    :desc "Open URL" "gx" #'browse-url-at-point)

(define-key evil-normal-state-map (kbd "C-a") 'evil-numbers/inc-at-pt)
(define-key evil-normal-state-map (kbd "C-x") 'evil-numbers/dec-at-pt)

(define-key evil-normal-state-map (kbd "C-l") 'evil-window-right)
(define-key evil-normal-state-map (kbd "C-h") 'evil-window-left)

(define-key evil-insert-state-map (kbd "C-l") 'evil-window-right)
(define-key evil-insert-state-map (kbd "C-h") 'evil-window-left)

(defun toggle-maximize-window ()
 "Maximize the window or bring back the previous layout."
  (interactive)
  (if (one-window-p)
    (winner-undo)
    (delete-other-windows)))

(define-key evil-normal-state-map (kbd "C-SPC") 'toggle-maximize-window)

(define-key evil-visual-state-map (kbd "K") 'drag-stuff-up)
(define-key evil-visual-state-map (kbd "J") 'drag-stuff-down)

(setq doom-theme 'doom-gruvbox-material) ; dark variant
(setq doom-font (font-spec :family "JetBrainsMono Nerd Font" :size 20 :height 1.0 :weight 'normal))
(setq truncate-lines 'nil)

(setq highlight-indent-guides-auto-odd-face-perc 50)
(setq highlight-indent-guides-auto-even-face-perc 50)
(setq highlight-indent-guides-auto-character-face-perc 50)
(setq highlight-indent-guides-responsive 'top)

(setq display-line-numbers-type 'relative)

(setq org-hide-emphasis-markers t)

(use-package! org-appear
  :defer t
  :config
  (setq org-appear-autolinks t))
(add-hook! org-mode :append #'org-appear-mode)

(use-package! company
  :defer t
  :custom-face
  (company-tooltip
   ((t (:family "JetBrainsMono Nerd Font" :height 1.0))))
  :config
        (define-key evil-insert-state-map (kbd "C-y") 'company-complete)
        (setq company-idle-delay 0.1)
        (setq company-tooltip-idle-delay 0.1)
        (add-to-list '+lsp-company-backends 'company-files)
        (set-company-backend! 'org-mode 'company-files 'company-capf)
)
(with-eval-after-load 'company
  ;; Unbind RET so it does not complete
  (define-key company-active-map (kbd "RET") nil)
  (define-key company-active-map (kbd "<return>") nil))

(map! :localleader
      :map LaTeX-mode-map
      :desc "Master"    "m" 'TeX-command-master)

(setq TeX-engine-alist '((default
                          "Tectonic"
                          "tectonic -X compile -f plain %T"
                          "tectonic -X watch"
                          nil)))

(setq LaTeX-command-style '(("" "%(latex)")))

(setq TeX-process-asynchronous t
      TeX-check-TeX nil
      TeX-engine 'default)

(add-hook! LaTeX-mode
(let ((tex-list (assoc "TeX" TeX-command-list))
      (latex-list (assoc "LaTeX" TeX-command-list)))
  (setf (cadr tex-list) "%(tex)"
        (cadr latex-list) "%l")))

(use-package! typopunct
  :defer t
  :config
  (typopunct-change-language 'french t))

(setq langtool-language-tool-jar (concat (getenv "HOME") "/bin/LanguageTool-5.2/languagetool-commandline.jar"))

(defun langtool-autoshow-detail-popup (overlays)
  (when (require 'popup nil t)
    ;; Do not interrupt current popup
    (unless (or popup-instances
                ;; suppress popup after type `C-g` .
                (memq last-command '(keyboard-quit)))
      (let ((msg (langtool-details-error-message overlays)))
        (popup-tip msg)))))
(setq langtool-autoshow-message-function
      'langtool-autoshow-detail-popup)

(use-package! yaml-mode
  :defer t
  :mode ("\\.ya?ml\\'" . yaml-mode))

(setq python-shell-interpreter "ipython"
    python-shell-interpreter-args "-i --simple-prompt")

(defvar python--pdb-breakpoint-string "breakpoint()")
(defun python-add-breakpoint ()
"Inserts a python breakpoint using `pdb'"
    (interactive)
    (back-to-indentation)
    ;; this preserves the correct indentation in case the line above
    ;; point is a nested block
    (split-line)
    (insert python--pdb-breakpoint-string)
    (python-set-debug-highlight))

(defun python-set-debug-highlight ()
    (interactive)
    (highlight-lines-matching-regexp "breakpoint[.]?" 'hi-red-b))

(defun python-add-debug-highlight ()
    "Adds a highlighter for use by `python--pdb-breakpoint-string'"
    (interactive)
    (python-set-debug-highlight))
    (add-hook 'python-ts-mode-hook 'python-add-debug-highlight)

(map! :localleader
      :map python-ts-mode-map
      :desc "Insert breakpoint"    "d" 'python-add-breakpoint)

(use-package! numpydoc
  :defer t
  :init
  (setq numpydoc-insertion-style 'yas))

(map! :localleader
      :map python-ts-mode-map
      :desc "Auto docstring"  "s" 'numpydoc-generate)

(setq +format-on-save-disabled-modes (add-to-list '+format-on-save-disabled-modes 'dockerfile-mode))

(setq lsp-go-use-gofumpt t)

(map! :leader
      (:prefix-map ("r" . "Harpoon")
       (:desc "Menu" "m" #'harpoon-quick-menu-hydra
        :desc "Add file" "a" #'harpoon-add-file
        :desc "Edit file" "r" #'harpoon-toggle-file
        :desc "Clear" "c" 'harpoon-clear)))

(map! :leader "1" 'harpoon-go-to-1)
(map! :leader "2" 'harpoon-go-to-2)
(map! :leader "3" 'harpoon-go-to-3)
(map! :leader "4" 'harpoon-go-to-4)
(map! :leader "5" 'harpoon-go-to-5)
(map! :leader "6" 'harpoon-go-to-6)
(map! :leader "7" 'harpoon-go-to-7)
(map! :leader "8" 'harpoon-go-to-8)
(map! :leader "9" 'harpoon-go-to-9)
(map! :leader "0" 'harpoon-go-to-10)

(setq org-export-use-babel t)
(setq org-directory "~/org/")
(setq org-latex-pdf-process '("tectonic %f"))
(setq org-export-in-background t)

(use-package! org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config
  (setq org-auto-tangle-default t))

(map! :leader
      :map org-mode-map
      :desc "Preview LaTeX" "mu" 'org-latex-preview)

(after! ox-latex
    (add-to-list 'org-latex-classes
                '("koma-article" "\\documentclass{scrartcl}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

    (add-to-list 'org-latex-classes
                '("koma-article-fr" "\\documentclass[french]{scrartcl}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

    (add-to-list 'org-latex-classes
                '("memoir-fr"
                "\\documentclass[a4paper,11pt,titlepage, twoside]{memoir}
                    \\usepackage[utf8]{inputenc}
                    \\usepackage[T1]{fontenc}
                    \\usepackage{fixltx2e}
                    \\usepackage{hyperref}
                    \\usepackage{mathpazo}
                    \\usepackage{color}
                    \\usepackage{enumerate}
                    \\definecolor{bg}{rgb}{0.95,0.95,0.95}
                    \\tolerance=1000
                    \\linespread{1.1}
                    \\hypersetup{pdfborder=0 0 0}"
                ("\\chapter{%s}" . "\\chapter*{%s}")
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

    (add-to-list 'org-latex-classes
                '("TMI"
                "\\documentclass[journal, web, twoside]{ieeecolor}"
                ("\\section{%s}" . "\\section*{%s}")
                ("\\subsection{%s}" . "\\subsection*{%s}")
                ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                ("\\paragraph{%s}" . "\\paragraph*{%s}")
                ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))

)

(use-package! ox-awesomecv
  :after org)

(defun my/org-inline-css-hook (exporter)
  "Insert custom inline css to automatically set the
background of code to whatever theme I'm using's background"
  (when (eq exporter 'html)
    (let* ((my-pre-bg (face-background 'default))
           (my-pre-fg (face-foreground 'default)))
      (setq
       org-html-head-extra
       (concat
        org-html-head-extra
        (format "<style type=\"text/css\">\n pre.src {background-color: %s; color: %s;}</style>\n"
                my-pre-bg my-pre-fg))))))

(add-hook 'org-export-before-processing-hook 'my/org-inline-css-hook)

(use-package! org-ref
  :defer t
    :init
        (setq org-ref-bibliography-notes "~/org/paper-notes/paper-notes.org"
            org-ref-default-bibliography "~/org/refs.bib"
            bibtex-completion-bibliography org-ref-default-bibliography
            bibtex-completion-notes-path "~/org/paper-notes/paper-notes.org"
            bibtex-completion-pdf-open-function
                (lambda (fpath)
                (call-process "zathura" nil 0 nil fpath))))

(map! :localleader
      :map org-mode-map
      :desc "Insert citation" "c" 'org-cite-insert)

(setq org-capture-templates
  (quote
   (("t" "todo" entry
     (file+headline "~/org/todo.org" "Tasks")
     "* TODO %U %?\n\n"
     :empty-lines-after 1)
    ("n" "note" entry
     (file+headline "~/org/notes.org" "Inbox")
     "* %U %? \n\n"
     :empty-lines-after 1))))

(after! counsel
  (setq counsel-rg-base-command "rg -M 240 --with-filename --no-heading --line-number --color never %s || true"))

(setq confirm-kill-emacs nil)

(setq ll/notmuch-default-query "tag:inbox AND not tag:deleted AND date:3months..")
(after! notmuch
    (setq notmuch-saved-searches
        `((:name "inbox"
        :query ,ll/notmuch-default-query
        :key "i"
        :search-type: tree)
        (:name "sent"
        :query "tag:sent AND not tag:deleted")
        (:name "gmail"
        :query "tag:Gmail/Inbox AND not tag:deleted"
        :key "g"
        :search-type: tree)
        (:name "personal"
        :query "tag:Gandi/Inbox AND not tag:deleted"
        :key "p"
        :search-type tree)))
    (setq +notmuch-mail-folder "~/.mail")
    (setq +notmuch-sync-backend 'mbsync)
    (setq +notmuch-home-function (lambda () (notmuch-search ll/notmuch-default-query)))
    
    (map! :leader
        :map notmuch-mode-map
        :desc "Saved searches" "mj" 'notmuch-jump-search))

(setq notmuch-always-prompt-for-sender t)
(setq user-full-name "Laurent Lejeune")
(setq user-mail-address "me@lejeunel.org")
(setq mail-host-address "lejeunel.org")

(after! notmuch
  (setq notmuch-show-part-action-list
        '(("Open with xdg-open" . (lambda (part) (start-process "xdg-open" nil "xdg-open" (notmuch-show-get-filename part))))
          ("View" . notmuch-show-view-part)
          ("Save" . notmuch-show-save-part))))

(setq mail-specify-envelope-from t)
(setq message-sendmail-envelope-from 'header)
(setq mail-envelope-from 'header)

(setq message-make-forward-subject-function 'message-forward-subject-fwd)

(defun my-notmuch-mua-empty-subject-check ()
  "Request confirmation before sending a message with empty subject"
  (when (and (null (message-field-value "Subject"))
             (not (y-or-n-p "Subject is empty, send anyway? ")))
    (error "Sending message cancelled: empty subject.")))
(add-hook 'message-send-hook 'my-notmuch-mua-empty-subject-check)

(after! vterm
  (set-popup-rule! "^.*vterm.*" :size 0.5 :vslot -4 :select t :quit nil :ttl 0 :side 'right))
(map! :leader
    :desc "Open vterm" "v" #'+vterm/here)

(general-create-definer comma-leader-def
  :states '(normal)
  :prefix ",")

;; Bind commands under the comma prefix
(comma-leader-def
    :desc "Switch workspace" "," #'+workspace/switch-to
    :desc "Rename workspace" "r" #'+workspace/rename
    :desc "Kill workspace" "k" #'+workspace/kill
    :desc "New named workspace" "n" #'+workspace/new-named)
